extensions/amp-story/1.0/amp-story-page.js

```diff
@@ -322,17 +326,35 @@ export class AmpStoryPage extends AMP.BaseElement {
     this.maybeConvertCtaLayerToPageOutlink_();
   }
 
-  /** @private */
-  maybeSetPreviewDuration_() {
-    if (this.storeService_.get(StateProperty.PREVIEW_STATE)) {
-      const videos = this.getAllVideos_();
-
-      const autoAdvanceAttr =
-        videos.length > 0
-          ? VIDEO_PREVIEW_AUTO_ADVANCE_DURATION
-          : DEFAULT_PREVIEW_AUTO_ADVANCE_DURATION;
-
-      this.element.setAttribute('auto-advance-after', autoAdvanceAttr);
+  /**
+   * Configures the page to auto advance using preview-specific durations.
+   * @private
+   */
+  setupAutoAdvanceForPreview_() {
+    const firstVideo = this.getFirstAmpVideo_();
+    const autoAdvanceDuration = firstVideo
+      ? VIDEO_PREVIEW_AUTO_ADVANCE_DURATION
+      : DEFAULT_PREVIEW_AUTO_ADVANCE_DURATION;
+    this.element.setAttribute('auto-advance-after', autoAdvanceDuration);
+
+    if (firstVideo) {
+      whenUpgradedToCustomElement(firstVideo)
+        .then(() => firstVideo.getImpl())
+        .then((videoImpl) => {
+          const loopVideoIfTooShort = (vidLength) => {
+            const videoEl = firstVideo.querySelector('video');
+            const tooShort = vidLength < VIDEO_PREVIEW_AUTO_ADVANCE_DURATION_S;
+            videoEl.loop ||= tooShort;
+          };
+
+          if (videoImpl.getDuration()) {
+            loopVideoIfTooShort(videoImpl.getDuration());
+          } else {
+            listenOnce(firstVideo, VideoEvents_Enum.LOADEDMETADATA, () => {
```

Comment:
I think `this.loadPromise(firstVideo)` might work for both scenarios so we don't split the code: `this.loadPromise(firstVideo).then(() => loopVideoIfTooShort())`

The function checks the readyState to resolve, and if it's not loaded, waits for the loadedmetadata event, so essentially the same as you're doing here but with a helper

Reply:
Good to know, thanks!